include/cuco/detail/open_addressing_ref_impl.cuh

```diff
@@ -657,20 +663,39 @@ class open_addressing_ref_impl {
    *
    * @return Result of this operation, i.e., success/continue/duplicate
    */
-  template <typename Predicate>
+  template <bool IsSet, typename Predicate>
   [[nodiscard]] __device__ constexpr insert_result packed_cas(value_type* slot,
                                                               value_type const& value,
                                                               Predicate const& predicate) noexcept
   {
-    auto old      = compare_and_swap(slot, this->empty_slot_sentinel_, value);
-    auto* old_ptr = reinterpret_cast<value_type*>(&old);
-    if (cuco::detail::bitwise_compare(*old_ptr, this->empty_slot_sentinel_)) {
+    auto old            = compare_and_swap(slot, this->empty_slot_sentinel_, value);
+    auto* old_ptr       = reinterpret_cast<value_type*>(&old);
+    auto const inserted = [&]() {
+      if constexpr (IsSet) {
+        // If it's a set implementation, compare the whole slot content
+        return cuco::detail::bitwise_compare(*old_ptr, this->empty_slot_sentinel_);
+      }
+      if constexpr (not IsSet) {
+        // If it's a map implementation, compare keys only
+        return cuco::detail::bitwise_compare(old_ptr->first, this->empty_slot_sentinel_.first);
+      }
+    }();
+    if (inserted) {
       return insert_result::SUCCESS;
     } else {
       // Shouldn't use `predicate` operator directly since it includes a redundant bitwise compare
-      return predicate.equal_to(*old_ptr, value) == detail::equal_result::EQUAL
-               ? insert_result::DUPLICATE
-               : insert_result::CONTINUE;
+      auto const res = [&]() {
+        if constexpr (IsSet) {
+          // If it's a set implementation, compare the whole slot content
+          return predicate.equal_to(*old_ptr, value);
+        }
+        if constexpr (not IsSet) {
+          // If it's a map implementation, compare keys only
+          return predicate.equal_to(old_ptr->first, value.first);
```

Comment:
The return type in both cases is the same, so I think we can just use `if constexpr {...} else {...}`.